Sources/Forum.php

```diff
@@ -353,78 +637,61 @@ protected function main(): array|string
 		}
 
-		// Make sure that our scheduled tasks have been running as intended
+		// Make sure that our scheduled tasks have been running as intended.
 		Config::checkCron();
+	}
 
-		// Is the forum in maintenance mode? (doesn't apply to administrators.)
-		if (!empty(Config::$maintenance) && !User::$me->allowedTo('admin_forum')) {
-			// You can only login.... otherwise, you're getting the "maintenance mode" display.
-			if (isset($_REQUEST['action']) && (in_array($_REQUEST['action'], ['login2', 'logintfa', 'logout']))) {
-				return self::$actions[$_REQUEST['action']][1];
-			}
-
-			// Don't even try it, sonny.
-			return __CLASS__ . '::inMaintenance';
-		}
-
-		// If guest access is off, a guest can only do one of the very few following actions.
-		if (empty(Config::$modSettings['allow_guestAccess']) && User::$me->is_guest && (!isset($_REQUEST['action']) || !in_array($_REQUEST['action'], self::$guest_access_actions))) {
-			User::$me->kickIfGuest(null, false);
-		} elseif (empty($_REQUEST['action'])) {
-			// Action and board are both empty... BoardIndex! Unless someone else wants to do something different.
+	/**
+	 * Resolves the appropriate action to execute based on the current request context.
+	 *
+	 * @return string|callable|false Returns one of the following:
+	 *  - A string representing a class implementing ActionInterface.
+	 *  - A callable string representing a static method (e.g., `'Class::method'`).
+	 */
+	protected function findAction(): string|callable|false
+	{
+		// If no action was supplied, is there an implied action?
+		if (empty($_REQUEST['action'])) {
+			// Action and board are both empty... BoardIndex!
 			if (empty(Board::$info->id) && empty(Topic::$topic_id)) {
+				// ... unless someone else wants to do something different.
 				if (!empty(Config::$modSettings['integrate_default_action'])) {
-					$defaultAction = explode(',', Config::$modSettings['integrate_default_action']);
+					$default_action = explode(',', Config::$modSettings['integrate_default_action'])[0];
 
-					// Sorry, only one default action is needed.
-					$defaultAction = $defaultAction[0];
-
-					$call = Utils::getCallable($defaultAction);
-
-					if (!empty($call)) {
-						return $call;
-					}
+					return is_a($default_action, ActionInterface::class, true) ? $default_action : Utils::getCallable($default_action);
 				}
 
-				// No default action huh? then go to our good old BoardIndex.
-				else {
-					return 'SMF\\Actions\\BoardIndex::call';
-				}
+				$_REQUEST['action'] = 'boardindex';
 			}
-
 			// Topic is empty, and action is empty.... MessageIndex!
 			elseif (empty(Topic::$topic_id)) {
-				return 'SMF\\Actions\\MessageIndex::call';
+				$_REQUEST['action'] = 'messageindex';
 			}
-
-			// Board is not empty... topic is not empty... action is empty.. Display!
+			// Board is not empty... topic is not empty... action is empty... Display!
 			else {
-				return 'SMF\\Actions\\Display::call';
+				$_REQUEST['action'] = 'display';
```

Comment:
Any reason to fix the request here? I'm going to change this to a parameter variable.

Reply:
Generally speaking, I was going for greater consistency. Given that BoardIndex, MessageIndex, and Display behave like actions in every way except that they usually aren't accessed via `?action=...`, it's odd to have them behave differently in this regard.

More specifically, I have been working for the last little while on implementing a robust routing system (a.k.a. pretty URLs), and as part of that it is helpful to have fully defined and consistent mapping between URL query parameters and the called ActionInterface instances.

Reply:
That said, if you are planning to ditch direct use of `$_REQUEST['action']` in favour of something better, by all means do so. As long as BoardIndex, MessageIndex, and Display work like every other action in your new system, that will be fine.

Reply:
I just want to  avoid writing to superglobals if at all possible.

Reply:
That makes sense. The best way to do that would be to replace all references to `$_REQUEST['action']` with something like `QueryString::get('action')`... which would definitely go outside of the scope of this PR. 

So, sure, revert those writes to `$_REQUEST['action']` for now, and I will come up with something to implement a getter (and maybe a setter) for QueryString in a different PR later.

Reply:
I already pushed a  commit that shunted it to a variable.